src/providers/oracle/qgsoracleprovider.cpp

```diff
@@ -269,6 +269,19 @@ bool QgsOracleProvider::exec( QSqlQuery &qry, QString sql, const QVariantList &a
       qry.addBindValue( arg );
     }
     res = qry.exec();
+    if ( !res && retry )
+    {
+      // If the connection has been closed because of timeout
+      // it will be great to restart it
+      // ORA-12170: TNS:Connect timeout occurred
+      if ( qry.lastError().number() == 12170 )
+      {
+        disconnectDb();
```

Comment:
Would this need to be done also if there is no retry happening? Reading the description of this pull request it sounds like.

Reply:
Does not build - `::exec` is static - `::disconnectDb()` is not.

Reply:
@jef-n if `::exec` is static, it's not possible to create a new connection to replace the old one in timeout.

Reply:
> @jef-n if `::exec` is static, it's not possible to create a new connection to replace the old one in timeout.

Not without more changes at least (ie. keeping a reference to the provider in the feature iterator etc).